File: Sen/Single_Number_III.py
class Solution(object):
    def singleNumber(self, nums): # linear time but not constant space
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        if not len(nums):
            return []
        rst = []
        d = {}
        for i in range(len(nums)):
            if nums[i] not in d:
                d[nums[i]] = 1
            else:
                d[nums[i]] += 1
        
        for item in d.items():
            if item[1] == 1:
                rst.append(item[0])
        return rst

File: xiaoming/1018/SingleNumberIII.java

public class SingleNumberIII {
	
	public int[] singleNumber(int[] nums) {
        int[] res = new int[2];
        res[0] = 0;
        res[1] = 0;
        for (int i = 0; i < nums.length; i++) {
            res[0] ^= nums[i];
        }
        int mask = (res[0] - 1) ^ (-1) & res[0];
        res[0] = 0;
        for (int i = 0; i < nums.length; i++) {
            if ((nums[i] & mask) == 0)
                res[0] ^= nums[i];
            else res[1] ^= nums[i];
        }
        return res;
    }

}
